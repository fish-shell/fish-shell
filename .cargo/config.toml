# This file is _not_ included in the tarballs for now
# Binary builds on Linux packaging infrastructure need to overwrite it to make `cargo vendor` work
# Releases and development builds made using OBS/Launchpad will _not_ reflect the contents of this
# file

[resolver]
# Make cargo 1.84+ respect MSRV (rust-version in Cargo.toml)
incompatible-rust-versions = "fallback"
