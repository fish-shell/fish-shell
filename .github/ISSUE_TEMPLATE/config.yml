blank_issues_enabled: true
contact_links:
  - name: Gitter chat
    url: https://gitter.im/fish-shell/fish-shell
    about: You can ask and answer questions here
