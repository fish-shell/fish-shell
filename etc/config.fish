# Put system-wide fish configuration entries here
# or in .fish files in conf.d/
# Files in conf.d can be overridden by the user
# by files with the same name in $XDG_CONFIG_HOME/fish/conf.d

# This file is run by all fish instances.
# To include configuration only for login shells, use
# if status is-login
#    ...
# end
# To include configuration only for interactive shells, use
# if status is-interactive
#   ...
# end
