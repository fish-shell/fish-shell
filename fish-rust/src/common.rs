use crate::ffi;
use crate::wchar_ext::WExt;
use crate::wchar_ffi::c_str;
use crate::wchar_ffi::{wstr, WCharFromFFI, WString};
use std::{ffi::c_uint, mem};

/// A scoped manager to save the current value of some variable, and optionally set it to a new
/// value. When dropped, it restores the variable to its old value.
///
/// This can be handy when there are multiple code paths to exit a block.
pub struct ScopedPush<'a, T> {
    var: &'a mut T,
    saved_value: Option<T>,
}

impl<'a, T> ScopedPush<'a, T> {
    pub fn new(var: &'a mut T, new_value: T) -> Self {
        let saved_value = mem::replace(var, new_value);

        Self {
            var,
            saved_value: Some(saved_value),
        }
    }

    pub fn restore(&mut self) {
        if let Some(saved_value) = self.saved_value.take() {
            *self.var = saved_value;
        }
    }
}

impl<'a, T> Drop for ScopedPush<'a, T> {
    fn drop(&mut self) {
        self.restore()
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum EscapeStringStyle {
    Script(EscapeFlags),
    Url,
    Var,
    Regex,
}

/// Flags for the [`escape_string()`] function. These are only applicable when the escape style is
/// [`EscapeStringStyle::Script`].
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub struct EscapeFlags {
    /// Do not escape special fish syntax characters like the semicolon. Only escape non-printable
    /// characters and backslashes.
    pub no_printables: bool,
    /// Do not try to use 'simplified' quoted escapes, and do not use empty quotes as the empty
    /// string.
    pub no_quoted: bool,
    /// Do not escape tildes.
    pub no_tilde: bool,
    /// Replace non-printable control characters with Unicode symbols.
    pub symbolic: bool,
}

/// Replace special characters with backslash escape sequences. Newline is replaced with `\n`, etc.
pub fn escape_string(s: &wstr, style: EscapeStringStyle) -> WString {
    let mut flags_int = 0;

    let style = match style {
        EscapeStringStyle::Script(flags) => {
            const ESCAPE_NO_PRINTABLES: c_uint = 1 << 0;
            const ESCAPE_NO_QUOTED: c_uint = 1 << 1;
            const ESCAPE_NO_TILDE: c_uint = 1 << 2;
            const ESCAPE_SYMBOLIC: c_uint = 1 << 3;

            if flags.no_printables {
                flags_int |= ESCAPE_NO_PRINTABLES;
            }
            if flags.no_quoted {
                flags_int |= ESCAPE_NO_QUOTED;
            }
            if flags.no_tilde {
                flags_int |= ESCAPE_NO_TILDE;
            }
            if flags.symbolic {
                flags_int |= ESCAPE_SYMBOLIC;
            }

            ffi::escape_string_style_t::STRING_STYLE_SCRIPT
        }
        EscapeStringStyle::Url => ffi::escape_string_style_t::STRING_STYLE_URL,
        EscapeStringStyle::Var => ffi::escape_string_style_t::STRING_STYLE_VAR,
        EscapeStringStyle::Regex => ffi::escape_string_style_t::STRING_STYLE_REGEX,
    };

    ffi::escape_string(c_str!(s), flags_int.into(), style).from_ffi()
}

/// Test if the string is a valid function name.
pub fn valid_func_name(name: &wstr) -> bool {
    if name.is_empty() {
        return false;
    };
    if name.char_at(0) == '-' {
        return false;
    };
    // A function name needs to be a valid path, so no / and no NULL.
    if name.find_char('/').is_some() {
        return false;
    };
    if name.find_char('\0').is_some() {
        return false;
    };
    true
}

pub const fn assert_send<T: Send>() {}

pub const fn assert_sync<T: Sync>() {}
