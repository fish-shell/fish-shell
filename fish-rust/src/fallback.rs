//! This file only contains fallback implementations of functions which have been found to be missing
//! or broken by the configuration scripts.
//!
//! Many of these functions are more or less broken and incomplete.

use crate::widecharwidth::{WcLookupTable, WcWidth};
use crate::{common::is_console_session, wchar::prelude::*};
use once_cell::sync::Lazy;
use std::cmp;
use std::sync::atomic::{AtomicI32, Ordering};
use std::{ffi::CString, mem, os::fd::RawFd};

/// Width of ambiguous East Asian characters and, as of TR11, all private-use characters.
/// 1 is the typical default, but we accept any non-negative override via `$fish_ambiguous_width`.
#[no_mangle]
pub static FISH_AMBIGUOUS_WIDTH: AtomicI32 = AtomicI32::new(1);

/// Width of emoji characters.
///
/// This must be configurable because the value changed between Unicode 8 and Unicode 9, `wcwidth()`
/// is emoji-unaware, and terminal emulators do different things.
///
/// See issues like #4539 and https://github.com/neovim/issues/4976 for how painful this is.
///
/// Valid values are 1, and 2. 1 is the typical emoji width used in Unicode 8 while some newer
/// terminals use a width of 2 since Unicode 9.
// For some reason, this is declared here and exposed here, but is set in `env_dispatch`.
#[no_mangle]
pub static FISH_EMOJI_WIDTH: AtomicI32 = AtomicI32::new(1);

static WC_LOOKUP_TABLE: Lazy<WcLookupTable> = Lazy::new(WcLookupTable::new);

/// A safe wrapper around the system `wcwidth()` function
pub fn wcwidth(c: char) -> i32 {
    extern "C" {
        pub fn wcwidth(c: libc::wchar_t) -> libc::c_int;
    }

    const _: () = assert!(mem::size_of::<libc::wchar_t>() >= mem::size_of::<char>());
    unsafe { wcwidth(c as libc::wchar_t) }
}

// Big hack to use our versions of wcswidth where we know them to be broken, which is
// EVERYWHERE (https://github.com/fish-shell/fish-shell/issues/2199)
pub fn fish_wcwidth(c: char) -> i32 {
    // The system version of wcwidth should accurately reflect the ability to represent characters
    // in the console session, but knows nothing about the capabilities of other terminal emulators
    // or ttys. Use it from the start only if we are logged in to the physical console.
    if is_console_session() {
        return wcwidth(c);
    }

    // Check for VS16 which selects emoji presentation. This "promotes" a character like U+2764
    // (width 1) to an emoji (probably width 2). So treat it as width 1 so the sums work. See #2652.
    // VS15 selects text presentation.
    let variation_selector_16 = '\u{FE0F}';
    let variation_selector_15 = '\u{FE0E}';
    if c == variation_selector_16 {
        return 1;
    } else if c == variation_selector_15 {
        return 0;
    }

    // Check for Emoji_Modifier property. Only the Fitzpatrick modifiers have this, in range
    // 1F3FB..1F3FF. This is a hack because such an emoji appearing on its own would be drawn as
    // width 2, but that's unlikely to be useful. See #8275.
    if ('\u{F3FB}'..='\u{1F3FF}').contains(&c) {
        return 0;
    }

    let width = WC_LOOKUP_TABLE.classify(c);
    match width {
        WcWidth::NonCharacter | WcWidth::NonPrint | WcWidth::Combining | WcWidth::Unassigned => {
            // Fall back to system wcwidth in this case.
            wcwidth(c)
        }
        WcWidth::Ambiguous | WcWidth::PrivateUse => {
            // TR11: "All private-use characters are by default classified as Ambiguous".
            FISH_AMBIGUOUS_WIDTH.load(Ordering::Relaxed)
        }
        WcWidth::One => 1,
        WcWidth::Two => 2,
        WcWidth::WidenedIn9 => FISH_EMOJI_WIDTH.load(Ordering::Relaxed),
    }
}

/// fish's internal versions of wcwidth and wcswidth, which can use an internal implementation if
/// the system one is busted.
pub fn fish_wcswidth(s: &wstr) -> i32 {
    let mut result = 0;
    for c in s.chars() {
        let w = fish_wcwidth(c);
        if w < 0 {
            return -1;
        }
        result += w;
    }
    result
}

// Replacement for mkostemp(str, O_CLOEXEC)
// This uses mkostemp if available,
// otherwise it uses mkstemp followed by fcntl
pub fn fish_mkstemp_cloexec(name_template: CString) -> (RawFd, CString) {
    let name = name_template.into_raw();
    #[cfg(not(target_os = "macos"))]
    let fd = {
        use libc::O_CLOEXEC;
        unsafe { libc::mkostemp(name, O_CLOEXEC) }
    };
    #[cfg(target_os = "macos")]
    let fd = {
        use libc::{FD_CLOEXEC, F_SETFD};
        let fd = unsafe { libc::mkstemp(name) };
        if fd != -1 {
            unsafe { libc::fcntl(fd, F_SETFD, FD_CLOEXEC) };
        }
        fd
    };
    (fd, unsafe { CString::from_raw(name) })
}

pub fn wcscasecmp(lhs: &wstr, rhs: &wstr) -> cmp::Ordering {
    use std::char::ToLowercase;
    use widestring::utfstr::CharsUtf32;

    /// This struct streams the underlying lowercase chars of a `UTF32String` without allocating.
    ///
    /// `char::to_lowercase()` returns an iterator of chars and we sometimes need to cmp the last
    /// char of one char's `to_lowercase()` with the first char of the other char's
    /// `to_lowercase()`. This makes that possible.
    struct ToLowerBuffer<'a> {
        current: ToLowercase,
        chars: CharsUtf32<'a>,
    }

    impl<'a> Iterator for ToLowerBuffer<'a> {
        type Item = char;

        fn next(&mut self) -> Option<Self::Item> {
            if let Some(c) = self.current.next() {
                return Some(c);
            }

            self.current = self.chars.next()?.to_lowercase();
            self.next()
        }
    }

    impl<'a> ToLowerBuffer<'a> {
        pub fn from(w: &'a wstr) -> Self {
            let mut empty = 'a'.to_lowercase();
            let _ = empty.next();
            debug_assert!(empty.next().is_none());
            let mut chars = w.chars();
            Self {
                current: chars.next().map(|c| c.to_lowercase()).unwrap_or(empty),
                chars,
            }
        }
    }

    let lhs = ToLowerBuffer::from(lhs);
    let rhs = ToLowerBuffer::from(rhs);
    lhs.cmp(rhs)
}

#[test]
fn test_wcscasecmp() {
    use std::cmp::Ordering;

    // Comparison with empty
    assert_eq!(wcscasecmp(L!("a"), L!("")), Ordering::Greater);
    assert_eq!(wcscasecmp(L!(""), L!("a")), Ordering::Less);
    assert_eq!(wcscasecmp(L!(""), L!("")), Ordering::Equal);

    // Basic comparison
    assert_eq!(wcscasecmp(L!("A"), L!("a")), Ordering::Equal);
    assert_eq!(wcscasecmp(L!("B"), L!("a")), Ordering::Greater);
    assert_eq!(wcscasecmp(L!("A"), L!("B")), Ordering::Less);

    // Multi-byte comparison
    assert_eq!(wcscasecmp(L!("İ"), L!("i\u{307}")), Ordering::Equal);
    assert_eq!(wcscasecmp(L!("ia"), L!("İa")), Ordering::Less);
}
