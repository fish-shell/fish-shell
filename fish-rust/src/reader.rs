pub fn reader_schedule_prompt_repaint() {
    todo!()
}

/// \return whether fish is currently unwinding the stack in preparation to exit.
pub fn fish_is_unwinding_for_exit() -> bool {
    todo!()
}
