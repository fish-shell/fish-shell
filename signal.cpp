/** \file signal.c

The library for various signal related issues

*/

#include "config.h"


#include <stdlib.h>
#include <stdio.h>
#include <sys/types.h>
#include <signal.h>
#include <dirent.h>
#include <unistd.h>
#include <errno.h>

#ifdef HAVE_SIGINFO_H
#include <siginfo.h>
#endif

#include "common.h"
#include "fallback.h"
#include "util.h"

#include "wutil.h"
#include "signal.h"
#include "event.h"
#include "reader.h"
#include "proc.h"


/**
   Struct describing an entry for the lookup table used to convert
   between signal names and signal ids, etc.
*/
struct lookup_entry
{
    /**
       Signal id
    */
    int signal;
    /**
       Signal name
    */
    const wchar_t *name;
    /**
       Signal description
    */
    const wchar_t *desc;
};

/**
   The number of signal blocks in place. Increased by signal_block, decreased by signal_unblock.
*/
static int block_count=0;


/**
   Lookup table used to convert between signal names and signal ids,
   etc.
*/
static const struct lookup_entry lookup[] =
{
#ifdef SIGHUP
    {
        SIGHUP,
        L"SIGHUP",
        N_(L"Terminal hung up")
    }
    ,
#endif
#ifdef SIGINT
    {
        SIGINT,
        L"SIGINT",
        N_(L"Quit request from job control (^C)")
    }
    ,
#endif
#ifdef SIGQUIT
    {
        SIGQUIT,
        L"SIGQUIT",
        N_(L"Quit request from job control with core dump (^\\)")
    }
    ,
#endif
#ifdef SIGILL
    {
        SIGILL,
        L"SIGILL",
        N_(L"Illegal instruction")
    }
    ,
#endif
#ifdef SIGTRAP
    {
        SIGTRAP,
        L"SIGTRAP",
        N_(L"Trace or breakpoint trap")
    }
    ,
#endif
#ifdef SIGABRT
    {
        SIGABRT,
        L"SIGABRT",
        N_(L"Abort")
    }
    ,
#endif
#ifdef SIGBUS
    {
        SIGBUS,
        L"SIGBUS",
        N_(L"Misaligned address error")
    }
    ,
#endif
#ifdef SIGFPE
    {
        SIGFPE,
        L"SIGFPE",
        N_(L"Floating point exception")
    }
    ,
#endif
#ifdef SIGKILL
    {
        SIGKILL,
        L"SIGKILL",
        N_(L"Forced quit")
    }
    ,
#endif
#ifdef SIGUSR1
    {
        SIGUSR1,
        L"SIGUSR1",
        N_(L"User defined signal 1")
    }
    ,
#endif
#ifdef SIGUSR2
    {
        SIGUSR2, L"SIGUSR2",
        N_(L"User defined signal 2")
    }
    ,
#endif
#ifdef SIGSEGV
    {
        SIGSEGV,
        L"SIGSEGV",
        N_(L"Address boundary error")
    }
    ,
#endif
#ifdef SIGPIPE
    {
        SIGPIPE,
        L"SIGPIPE",
        N_(L"Broken pipe")
    }
    ,
#endif
#ifdef SIGALRM
    {
        SIGALRM,
        L"SIGALRM",
        N_(L"Timer expired")
    }
    ,
#endif
#ifdef SIGTERM
    {
        SIGTERM,
        L"SIGTERM",
        N_(L"Polite quit request")
    }
    ,
#endif
#ifdef SIGCHLD
    {
        SIGCHLD,
        L"SIGCHLD",
        N_(L"Child process status changed")
    }
    ,
#endif
#ifdef SIGCONT
    {
        SIGCONT,
        L"SIGCONT",
        N_(L"Continue previously stopped process")
    }
    ,
#endif
#ifdef SIGSTOP
    {
        SIGSTOP,
        L"SIGSTOP",
        N_(L"Forced stop")
    }
    ,
#endif
#ifdef SIGTSTP
    {
        SIGTSTP,
        L"SIGTSTP",
        N_(L"Stop request from job control (^Z)")
    }
    ,
#endif
#ifdef SIGTTIN
    {
        SIGTTIN,
        L"SIGTTIN",
        N_(L"Stop from terminal input")
    }
    ,
#endif
#ifdef SIGTTOU
    {
        SIGTTOU,
        L"SIGTTOU",
        N_(L"Stop from terminal output")
    }
    ,
#endif
#ifdef SIGURG
    {
        SIGURG,
        L"SIGURG",
        N_(L"Urgent socket condition")
    }
    ,
#endif
#ifdef SIGXCPU
    {
        SIGXCPU,
        L"SIGXCPU",
        N_(L"CPU time limit exceeded")
    }
    ,
#endif
#ifdef SIGXFSZ
    {
        SIGXFSZ,
        L"SIGXFSZ",
        N_(L"File size limit exceeded")
    }
    ,
#endif
#ifdef SIGVTALRM
    {
        SIGVTALRM,
        L"SIGVTALRM",
        N_(L"Virtual timer expired")
    }
    ,
#endif
#ifdef SIGPROF
    {
        SIGPROF,
        L"SIGPROF",
        N_(L"Profiling timer expired")
    }
    ,
#endif
#ifdef SIGWINCH
    {
        SIGWINCH,
        L"SIGWINCH",
        N_(L"Window size change")
    }
    ,
#endif
#ifdef SIGWIND
    {
        SIGWIND,
        L"SIGWIND",
        N_(L"Window size change")
    }
    ,
#endif
#ifdef SIGIO
    {
        SIGIO,
        L"SIGIO",
        N_(L"I/O on asynchronous file descriptor is possible")
    }
    ,
#endif
#ifdef SIGPWR
    {
        SIGPWR,
        L"SIGPWR",
        N_(L"Power failure")
    }
    ,
#endif
#ifdef SIGSYS
    {
        SIGSYS,
        L"SIGSYS",
        N_(L"Bad system call")
    }
    ,
#endif
#ifdef SIGINFO
    {
        SIGINFO,
        L"SIGINFO",
        N_(L"Information request")
    }
    ,
#endif
#ifdef SIGSTKFLT
    {
        SIGSTKFLT,
        L"SISTKFLT",
        N_(L"Stack fault")
    }
    ,
#endif
#ifdef SIGEMT
    {
        SIGEMT,
        L"SIGEMT",
        N_(L"Emulator trap")
    }
    ,
#endif
#ifdef SIGIOT
    {
        SIGIOT,
        L"SIGIOT",
        N_(L"Abort (Alias for SIGABRT)")
    }
    ,
#endif
#ifdef SIGUNUSED
    {
        SIGUNUSED,
        L"SIGUNUSED",
        N_(L"Unused signal")
    }
    ,
#endif
    {
        0,
        0,
        0
    }
}
;


/**
   Test if \c name is a string describing the signal named \c canonical.
*/
static int match_signal_name(const wchar_t *canonical,
                             const wchar_t *name)
{
    if (wcsncasecmp(name, L"sig", 3)==0)
        name +=3;

    return wcscasecmp(canonical+3,name) == 0;
}


int wcs2sig(const wchar_t *str)
{
    int i;
    wchar_t *end=0;

    for (i=0; lookup[i].desc ; i++)
    {
        if (match_signal_name(lookup[i].name, str))
        {
            return lookup[i].signal;
        }
    }
    errno=0;
    int res = fish_wcstoi(str, &end, 10);
    if (!errno && res>=0 && !*end)
        return res;

    return -1;
}


const wchar_t *sig2wcs(int sig)
{
    int i;

    for (i=0; lookup[i].desc ; i++)
    {
        if (lookup[i].signal == sig)
        {
            return lookup[i].name;
        }
    }

    return _(L"Unknown");
}

const wchar_t *signal_get_desc(int sig)
{
    int i;

    for (i=0; lookup[i].desc ; i++)
    {
        if (lookup[i].signal == sig)
        {
            return _(lookup[i].desc);
        }
    }

    return _(L"Unknown");
}

/**
   Standard signal handler
*/
static void default_handler(int signal, siginfo_t *info, void *context)
{
    if (event_is_signal_observed(signal))
    {
        event_fire_signal(signal);
    }
}

/**
   Respond to a winch signal by checking the terminal size
*/
static void handle_winch(int sig, siginfo_t *info, void *context)
{
    common_handle_winch(sig);
    default_handler(sig, 0, 0);
}

/**
   Respond to a hup signal by exiting, unless it is caught by a
   shellscript function, in which case we do nothing.
*/
static void handle_hup(int sig, siginfo_t *info, void *context)
{
    if (event_is_signal_observed(SIGHUP))
    {
        default_handler(sig, 0, 0);
    }
    else
    {
        reader_exit(1, 1);
    }
}

/** Handle sigterm. The only thing we do is restore the front process ID, then die. */
static void handle_term(int sig, siginfo_t *info, void *context)
{
    restore_term_foreground_process_group();
    signal(SIGTERM, SIG_DFL);
    raise(SIGTERM);
}

/**
   Interactive mode ^C handler. Respond to int signal by setting
   interrupted-flag and stopping all loops and conditionals.
*/
static void handle_int(int sig, siginfo_t *info, void *context)
{
    reader_handle_int(sig);
    default_handler(sig, info, context);
}

/**
   sigchld handler. Does notification and calls the handler in proc.c
*/
static void handle_chld(int sig, siginfo_t *info, void *context)
{
    job_handle_signal(sig, info, context);
    default_handler(sig, info, context);
}

void signal_reset_handlers()
{
    int i;

    struct sigaction act;
    sigemptyset(& act.sa_mask);
    act.sa_flags=0;
    act.sa_handler=SIG_DFL;

    for (i=0; lookup[i].desc ; i++)
    {
        sigaction(lookup[i].signal, &act, 0);
    }
}


/**
   Sets appropriate signal handlers.
*/
void signal_set_handlers()
{
    struct sigaction act;

    if (get_is_interactive() == -1)
        return;

    sigemptyset(& act.sa_mask);
    act.sa_flags=SA_SIGINFO;
    act.sa_sigaction = &default_handler;

    /*
      First reset everything to a use default_handler, a function
      whose sole action is to fire of an event
    */
    sigaction(SIGINT, &act, 0);
    sigaction(SIGQUIT, &act, 0);
    sigaction(SIGTSTP, &act, 0);
    sigaction(SIGTTIN, &act, 0);
    sigaction(SIGTTOU, &act, 0);
    sigaction(SIGCHLD, &act, 0);

    /*
      Ignore sigpipe, it is generated if fishd dies, but we can
      recover.
    */
    sigaction(SIGPIPE, &act, 0);

    if (get_is_interactive())
    {
        /*
           Interactive mode. Ignore interactive signals.  We are a
           shell, we know whats best for the user. ;-)
        */

        act.sa_handler=SIG_IGN;

        sigaction(SIGINT, &act, 0);
        sigaction(SIGQUIT, &act, 0);
        sigaction(SIGTSTP, &act, 0);
        sigaction(SIGTTIN, &act, 0);
        sigaction(SIGTTOU, &act, 0);

        act.sa_sigaction = &handle_int;
        act.sa_flags = SA_SIGINFO;
        if (sigaction(SIGINT, &act, 0))
        {
            wperror(L"sigaction");
            FATAL_EXIT();
        }

        act.sa_sigaction = &handle_chld;
        act.sa_flags = SA_SIGINFO;
        if (sigaction(SIGCHLD, &act, 0))
        {
            wperror(L"sigaction");
            FATAL_EXIT();
        }

#ifdef SIGWINCH
        act.sa_flags = SA_SIGINFO;
        act.sa_sigaction= &handle_winch;
        if (sigaction(SIGWINCH, &act, 0))
        {
            wperror(L"sigaction");
            FATAL_EXIT();
        }
#endif

        act.sa_flags = SA_SIGINFO;
        act.sa_sigaction= &handle_hup;
        if (sigaction(SIGHUP, &act, 0))
        {
            wperror(L"sigaction");
            FATAL_EXIT();
        }

        // SIGTERM restores the terminal controlling process before dying
        act.sa_flags = SA_SIGINFO;
        act.sa_sigaction= &handle_term;
        if (sigaction(SIGTERM, &act, 0))
        {
            wperror(L"sigaction");
            FATAL_EXIT();
        }

    }
    else
    {
        /*
          Non-interactive. Ignore interrupt, check exit status of
          processes to determine result instead.
        */
        act.sa_handler=SIG_IGN;

        sigaction(SIGINT, &act, 0);
        sigaction(SIGQUIT, &act, 0);

        act.sa_handler=SIG_DFL;

        act.sa_sigaction = &handle_chld;
        act.sa_flags = SA_SIGINFO;
        if (sigaction(SIGCHLD, &act, 0))
        {
            wperror(L"sigaction");
            exit_without_destructors(1);
        }
    }

}

void signal_handle(int sig, int do_handle)
{
    struct sigaction act;

    /*
      These should always be handled
    */
    if ((sig == SIGINT) ||
            (sig == SIGQUIT) ||
            (sig == SIGTSTP) ||
            (sig == SIGTTIN) ||
            (sig == SIGTTOU) ||
            (sig == SIGCHLD))
        return;

    sigemptyset(&act.sa_mask);
    if (do_handle)
    {
        act.sa_flags = SA_SIGINFO;
        act.sa_sigaction = &default_handler;
    }
    else
    {
        act.sa_flags = 0;
        act.sa_handler = SIG_DFL;
    }

    sigaction(sig, &act, 0);
}

void get_signals_with_handlers(sigset_t *set)
{
    sigemptyset(set);
    for (int i=0; lookup[i].desc ; i++)
    {
        struct sigaction act = {};
        sigaction(lookup[i].signal, NULL, &act);
        if (act.sa_handler != SIG_DFL)
            sigaddset(set, lookup[i].signal);
    }
}

void signal_block()
{
    ASSERT_IS_MAIN_THREAD();
    sigset_t chldset;

    if (!block_count)
    {
        sigfillset(&chldset);
        VOMIT_ON_FAILURE(pthread_sigmask(SIG_BLOCK, &chldset, NULL));
    }

    block_count++;
//	debug( 0, L"signal block level increased to %d", block_count );
}

void signal_unblock()
{
    ASSERT_IS_MAIN_THREAD();
    sigset_t chldset;

    block_count--;

    if (block_count < 0)
    {
        debug(0, _(L"Signal block mismatch"));
        bugreport();
        FATAL_EXIT();
    }

    if (!block_count)
    {
        sigfillset(&chldset);
        VOMIT_ON_FAILURE(pthread_sigmask(SIG_UNBLOCK, &chldset, 0));
    }
//	debug( 0, L"signal block level decreased to %d", block_count );
}

bool signal_is_blocked()
{
    return !!block_count;
}

