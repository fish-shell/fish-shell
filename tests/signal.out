ALRM received
command false:
command true:
command false | true:
This is the process whose exit event shuld be blocked
This should come before the event handler
Now event handler should have run
PROCESS_EXIT 0
